src/main/java/io/pravega/leaderelection/LeaderElection.java

```diff
@@ -0,0 +1,493 @@
+                        state.vectorTime,
+                        newLeader);
+        }
+    }
+
+    @RequiredArgsConstructor
+    private static class AddMember extends HeartbeatUpdate {
+        private static final long serialVersionUID = 1L;
+        private final String name;
+        private final long time;
+
+        @Override
+        public LiveInstances applyTo(LiveInstances state, Revision newRevision) {
+            Map<String, InstanceInfo> tempInstances = new HashMap<>(state.liveInstances);
+            tempInstances.put(name, new InstanceInfo(time, 1, INITIAL_TIMEOUT));
+            return new LiveInstances(state.scopedStreamName,
+                    newRevision,
+                    Collections.unmodifiableMap(tempInstances),
+                    state.vectorTime,
+                    state.leaderName);
+        }
+    }
+
+    /**
+     * Send new heartbeat to Pravega through the state Synchronizer.
+     * It will do three things:
+     * 1. update timestamp and timeout.
+     * 2. remove all dead instances.
+     * 3. If leader dead or not exist, select a new one.
+     */
+    private void sendHeartbeat() {
+        stateSync.updateState((state, updates) -> {
+            long vectorTime = state.getVectorTime() + 1;
+            InstanceInfo instanceInfo = state.liveInstances.get(instanceId);
+            long newTimes = instanceInfo.times + 1;
+            double newTimeout = (instanceInfo.timeout * instanceInfo.times +
+                    (vectorTime - instanceInfo.timestamp) * 1.0 /
+                            state.liveInstances.size()) / newTimes;
+
+            updates.add(new HeartBeat(instanceId, new InstanceInfo(vectorTime, newTimes, newTimeout)));
+
+            for (String id : state.findInstancesThatWillDieBy(vectorTime)) {
+                if (!id.equals(instanceId)) {
+                    updates.add(new DeclareDead(id));
+                    if (id.equals(state.leaderName)) {
+                        String newLeader = state.liveInstances.entrySet()
+                                .stream().max(LiveInstances::compare).get().getKey();
+                        updates.add(new SetLeader(newLeader));
+                    }
+                }
+            }
+            // for initial state or other states that leader doesn't exist
+            if (state.leaderName == null) {
+                String newLeader = state.liveInstances.entrySet()
+                        .stream().max(LiveInstances::compare).get().getKey();
+                updates.add(new SetLeader(newLeader));
+            }
+        });
+    }
+
+    private class HeartBeater implements Runnable {
+        @Override
+        public void run() {
+            try {
+                stateSync.fetchUpdates();
+                notifyListener();
+                sendHeartbeat();
+                // when leader changes, notify to all.
+                if (leaderName == null || !leaderName.equals(stateSync.getState().leaderName)) {
```

Comment:
This is comparing and changing leaderName without any threadsafe check. You probably need a synchronized block. My advice would be to make a tiny private class that just manages the leader variable. that way the outer class won't have to think about it.

Reply:
The leadername is a local copy of the leadername in state synchronizer. I think there is no need to do threadsafe check. only the heartbeater thread can check and change the local leadername.